lib/assets/javascripts/_modules/search.js

```diff
@@ -0,0 +1,171 @@
+    var $searchResults;
+    var $searchResultsTitle;
+    var $searchResultsWrapper;
+    var $searchResultsClose;
+    var results;
+    var maxSearchEntries = 10;
+
+    this.start = function start($element) {
+      $searchForm = $element.find('form');
+      $searchInput = $element.find('#search');
+      $searchResultsWrapper = $element.find('.search-results')
+      $searchResults = $searchResultsWrapper.find('.search-results__content');
+      $searchResultsTitle = $searchResultsWrapper.find('.search-results__title');
+      $searchResultsClose = $searchResultsWrapper.find('.search-results__close');
+      downloadSearchIndex();
+      attach();
+    };
+
+    function downloadSearchIndex() {
+      updateTitle('Loading search index')
+      $.ajax({
+        url: '/search.json',
+        cache: true,
+        method: 'GET',
+        success: function(data) {
+          lunrData = data;
+          lunrIndex = lunr.Index.load(lunrData.index);
+          $(document).trigger('lunrIndexLoaded');
+        }
+      });
+    }
+
+    function attach() {
+      // Search functionality on search text input
+      $searchInput.on('input', function (e) {
+        e.preventDefault();
+        var query = $(this).val();
+        search(query);
+      });
+
+      // Set focus on the first search result instead of submiting the search  
+      // form to Google
+      $searchForm.on('submit', function(e) {
+        e.preventDefault();
+        $searchResults.find('.search-result__title a').first().focus();
+      });
+
+      // Closing the search results, move focus back to the search input
+      $searchResultsClose.on('click', function(e) {
+        e.preventDefault();
+        $searchInput.focus();
+        hideResults();
+      });
+    }
+
+    function searchIndex(query) {
+      var results = [];
+      lunrIndex.search(query).forEach( function (item, index) {
+        if ( index < maxSearchEntries ) {
+          results.push(lunrData.docs[item.ref]);
+        }
+      });
+
+      return results;
+    }
+
+    function search(query) {
+      if(query === '') {
+        hideResults();
+        return;
+      }
+      showResults();
+      // The index has not been downloaded yet, exit early and wait.
+      if(!lunrIndex) {
+        $(document).on('lunrIndexLoaded', function() {
+          search(query);
+        });
+        return;
+      }
+      results = searchIndex(query);
+      renderResults(query);
+      updateTitle();
+    }
+
+    function renderResults(query) {
+      var output = '';
+      if (results.length == 0) {
+        output += '<p>Nothing found</p>';
+      }
+      output += '<ul>';
+      for(var index in results) {
+        var result = results[index];
+        var content = processContent(result.content, query);
+        output += '<li class="search-result">';
+        output += '<h3 class="search-result__title">';
+        output += '<a href="' + result.url + '">';
+        output += result.title;
```

Comment:
This could inject HTML into the page. It's likely that at some point it will contain characters that require HTML encoding like &<> which would break the markup of the page. Can we use a templating language instead of string concatenation?

Reply:
I started looking into potential templating systems but when writing the unit test I realised that the index we generate with lunrjs doesn't include any HTML.

I'm not against adding a templating system, but I'd rather not risk delaying this merge on a potential bike shed. How about we open a new issue for that?